### PR TITLE
Buffering readonly files

### PR DESCRIPTION
**Pull request by Razz4780**
Issue #2069 

Moved file ops to its own proxy inside intproxy. The logic around files is getting to complex to be kept in `SimpleProxy` (which was supposed to just pass messages back and forth).

Haven't tested the solution with jdbc yet

### CODE REVIEW

File: mirrord/intproxy/src/proxies/files.rs [hunk 1]
```diff
@@ -0,0 +1,1345 @@
...
+                    .send(ClientMessage::FileRequest(FileRequest::OpenRelative(open)))
+                    .await;
+            }
+
+            // Try to use local buffer if possible.
+            FileRequest::Read(read) => match self.files_data.get_mut(&read.remote_fd) {
+                // File is buffered.
+                Some(data) => {
+                    let from_buffer = data.read_from_buffer(read.buffer_size, data.fd_position);
+                    if let Some(from_buffer) = from_buffer {
```
Comment (meowjesty) [inline mirrord/intproxy/src/proxies/files.rs:307]: We read from the buffer, if the user asked to read more than what the buffer has, then we go into the `None` case which will re-fill the buffer with new data?

File: mirrord/intproxy/src/proxies/files.rs [hunk 2]
```diff
@@ -0,0 +1,1345 @@
...
+                        let bytes = from_buffer.to_vec();
+                        data.fd_position += read.buffer_size;
```
Comment (meowjesty) [inline mirrord/intproxy/src/proxies/files.rs:309]: `fd_position`? Do you mean that this is the cursor position in this buffer for the `fd` of the file we are reading?

File: mirrord/intproxy/src/request_queue.rs [hunk 3]
```diff
@@ -14,44 +14,59 @@
...
-        self.inner.push_back((message_id, layer_id));
+        self.inner
+            .push_back((message_id, layer_id, Default::default()));
+    }
+}
+
+impl<T: fmt::Debug> RequestQueue<T> {
+    /// Retrieve and remove a request from the front of this queue.
+    #[tracing::instrument(level = Level::TRACE)]
+    pub fn get(&mut self) -> Option<(MessageId, LayerId)> {
```
Comment (meowjesty) [inline mirrord/intproxy/src/request_queue.rs:56]: I'll forever resent that this is called `get` when we have `&mut self` going on. Shouldn't we change this name to `pop_front`, `pop`, `remove`?

File: mirrord/intproxy/src/request_queue.rs [hunk 4]
```diff
@@ -14,44 +14,59 @@
...
+    #[tracing::instrument(level = Level::TRACE)]
+    pub fn insert_with_data(&mut self, message_id: MessageId, layer_id: LayerId, data: T) {
+        self.inner.push_back((message_id, layer_id, data));
     }
 
     /// Retrieve and remove a request from the front of this queue.
     #[tracing::instrument(level = Level::TRACE)]
-    pub fn get(&mut self) -> Result<(MessageId, LayerId), RequestQueueEmpty> {
-        self.inner.pop_front().ok_or(RequestQueueEmpty)
+    pub fn get_with_data(&mut self) -> Option<(MessageId, LayerId, T)> {
```
Comment (meowjesty) [inline mirrord/intproxy/src/request_queue.rs:69]: Same here, feels weird to call this `get_`.

**Review by meowjesty — CHANGES_REQUESTED** [review]
I mostly have some questions to make sure I'm understanding some refactored stuff correctly.

File: mirrord/intproxy/src/proxies/files.rs (hunk 1 above)
Comment (Razz4780) [reply mirrord/intproxy/src/proxies/files.rs:307]: Yeah, `None` case will send a `ReadLimited` request to the agent for the next chunk

File: mirrord/intproxy/src/request_queue.rs (hunk 3 above)
Comment (Razz4780) [reply mirrord/intproxy/src/request_queue.rs:56]: Changed to `pop_front[_with_data]`, `push_back[_with_data]`

File: mirrord/intproxy/src/proxies/files.rs (hunk 2 above)
Comment (Razz4780) [reply mirrord/intproxy/src/proxies/files.rs:312]: I mean position of the file descriptor in the file, simulated in here. Normally fd position of remote files is handled in the agent - the agent opens a raw OS file descriptor, position of which is managed by the kernel. Following this approach for buffered files was a bit problematic. So I moved managing fd position to the intproxy (see `File buffering` section of `FilesProxy` doc)

File: mirrord/intproxy/src/request_queue.rs (hunk 3 above)
Comment (meowjesty) [reply mirrord/intproxy/src/request_queue.rs]: Are we objective-c?

**Review by meowjesty — APPROVED** [review]
:+1:

**Pull request merged**